ff-net/nn/Sequential.js

```diff
@@ -199,41 +197,46 @@ class Sequential {
   }
 
   train(args) {
-    // TODO decouple data from canvas
-    const dataCanvas = args.dataCanvas;
+    const dataPoints = args.dataPoints;
+    if (dataPoints == null) throw new Error("dataPoints required");
     const lr = args.lr;
-    const regularization = args.regularization;
-    const iters = args.iters;
+    if (lr == null) throw new Error("lr required");
+    const regularization = args.regularization ?? 0.0;
+    const iters = args.iters ?? 1;
 
     let regularizationLoss, dataLoss;
 
     const inputNeuronGroup = this.getInputNeuronGroup();
     const outputNeuronGroup = this.getOutputNeuronGroup();
+
     for (let i = 0; i < iters; i++) {
       dataLoss = 0;
-      dataCanvas.dataPoints.forEach((dataPoint) => {
+      // TODO batch mode?
+      dataPoints.forEach((dataPoint) => {
+        // forward
         // TODO generalize, do not assume 2D input
         inputNeuronGroup.neurons[0].activation = dataPoint.x;
         inputNeuronGroup.neurons[1].activation = dataPoint.y;
         this.forward();
-        
         const neuron = outputNeuronGroup.neurons[0];
         const output = neuron.activation;
         const d = dataPoint.label - output;
-        // forwardData
+        // TODO implement forwardData
         dataLoss += 0.5 * d * d;
-        
-        this.zeroGrad();
-        neuron.activationGrad = -d;
-
+        // TODO avoid computing regularization for every data point
         regularizationLoss = this.forwardRegularization({
           regularization: regularization
         });
+        
+        // backward
+        this.zeroGrad();
+        neuron.activationGrad = -d;
         this.backward();
         this.backwardRegularization({
           regularization: regularization
         });
-
+        
+        // optim
```

Comment:
This implementation is good enough for the demo, but in a general-purpose AD system we'd have additional differentiable nodes for the data loss and regularization loss. Since most of the functionality in our case (forward, backward, optim) is implemented in `Sequential`, we basically need to define two different forward and backward passes, one for data loss and one for regularization loss.